src/Features/InteriorSunShadows.cpp

```diff
@@ -0,0 +1,172 @@
+{
+	auto* singleton = GetSingleton();
+	const auto cell = globals::game::tes->interiorCell;
+	auto* passedJobArrays = &jobArrays;
+
+	if (!cell) {
+		singleton->ClearArrays();
+	} else {
+		const auto portalGraph = cell->GetRuntimeData().loadedData->portalGraph;
+		if (singleton->isInteriorWithSun && portalGraph) {
+			singleton->PopulateReplacementJobArrays(cell, portalGraph, dirLight, jobArrays);
+			passedJobArrays = &singleton->replacementJobArrays;
+		}
+	}
+
+	func(dirLight, *passedJobArrays, nodes);
+}
+
+void InteriorSunShadows::ClearArrays()
+{
+	if (arraysCleared)
+		return;
+
+	currentCellRoomsAndPortals.clear();
+
+	for (auto& jobArray : replacementJobArrays)
+		jobArray.clear();
+
+	arraysCleared = true;
+}
+
+namespace RE
+{
+	class BSMultiBoundRoom : public NiNode
+	{};
+}
+
+void InteriorSunShadows::PopulateReplacementJobArrays(RE::TESObjectCELL* cell, const RE::NiPointer<RE::BSPortalGraph>& portalGraph, const RE::BSShadowDirectionalLight* dirLight, RE::BSTArray<RE::BSTArray<RE::NiPointer<RE::NiAVObject>>>& jobArrays)
+{
+	if (cell != currentCell) {
+		InitialiseOnNewCell(portalGraph);
+		currentCell = cell;
+	}
+
+	const auto jobArraySize = jobArrays.size();
+
+	if (replacementJobArrays.size() != jobArraySize)
+		replacementJobArrays.resize(jobArraySize);
+
+	for (auto& jobArray : replacementJobArrays)
+		jobArray.clear();
+
+	addedSet.clear();
+
+	// Copy the original job arrays contents into the replacement job arrays
+	uint32_t count = 0;
+	for (uint32_t i = 0; i < jobArraySize; ++i) {
+		for (const auto& object : jobArrays[i]) {
+			replacementJobArrays[i].push_back(object);
+			addedSet.insert(object.get());
+			count++;
+		}
+	}
+
+	const auto playerPos = RE::PlayerCharacter::GetSingleton()->GetPosition();
+	auto lightDir = -dirLight->GetShadowDirectionalLightRuntimeData().lightDirection;
+	lightDir.Unitize();
+
+	// Add extra rooms and portals that are in the direction of the sun
+	for (const auto& object : currentCellRoomsAndPortals) {
+		if (addedSet.find(object.get()) != addedSet.end() || !IsInSunDirectionAndWithinShadowDistance(object, lightDir, playerPos))
+			continue;
+
+		addedSet.insert(object.get());
+		replacementJobArrays[count++ % jobArraySize].push_back(object);
+	}
+
+	arraysCleared = false;
+}
+
+void InteriorSunShadows::InitialiseOnNewCell(const RE::NiPointer<RE::BSPortalGraph>& portalGraph)
+{
+	currentCellRoomsAndPortals.clear();
+
+	if (const auto portalSharedNode = portalGraph->portalSharedNode) {
+		for (const auto room : portalGraph->rooms)
+			currentCellRoomsAndPortals.push_back(room);
+
+		for (auto child : portalGraph->portalSharedNode->GetChildren())
+			currentCellRoomsAndPortals.push_back(child);
+	}
+}
+
+bool InteriorSunShadows::IsInSunDirectionAndWithinShadowDistance(const RE::NiPointer<RE::NiAVObject>& object, const RE::NiPoint3& lightDir, const RE::NiPoint3& playerPos) const
+{
+	const float radius = object->worldBound.radius;
```

Comment:
No chance object is nullptr?

Reply:
These objects are the rooms/portals in the current cell and should never be null or something wrong has happened. I think I'd prefer it to crash here if that were the case.